File: src/db-gwent-server-scala/src/main/scala/entities/Turn.scala
package entities

case class Turn(players: List[PlayerId], var next: PlayerId):
  def run(playerId: PlayerId, f: () => Boolean): Boolean =
    canPlay(playerId) && f() && setPlayed(playerId)

  private def canPlay(playerId: PlayerId): Boolean =
    next == playerId

  private def setPlayed(playerId: PlayerId): Boolean =
    next = playerId; true

File: src/db-gwent-server-kotlin/src/main/kotlin/com/entities/Turn.kt
package com.entities

class Turn(val players: List<PlayerId>, var next: PlayerId = players.first()) {
    fun run(playerId: PlayerId, f: () -> Boolean): Boolean =
        canPlay(playerId) && f() && setPlayed(playerId)

    private fun canPlay(playerId: PlayerId): Boolean =
        next == playerId

    private fun setPlayed(playerId: PlayerId): Boolean {
        next = playerId
        return true
    }
}
